### ISSUE TITLE
Discovery and concensus: the long view of actor model architectures.

### ISSUE THREAD

**Issue by goodboy** [goodboy/tractor#184]
This is a draft issue that I hope gets filled out more and more by lurkers over time.

The road map has the following:
- custom cross-host supervisors
- mult-host service architecture tooling
- support for planet scale msg patterns and systems

But in order to do these things we need both a [service discovery system](https://en.wikipedia.org/wiki/Service_discovery) (presumably by relying on a (set of) protocols) **and**, a modern [concensus system](https://en.wikipedia.org/wiki/Consensus_(computer_science)) which these days is usually in the form of either a centralized algo such as [raft](https://en.wikipedia.org/wiki/Raft_(algorithm)) or, the thoroughly more *pumped*  [distributed equivalents](https://en.wikipedia.org/wiki/Quantum_Byzantine_agreement). 

Here are some resources to start poking at for research and tinkering:
- `RAFT` pumps:
  - this cool [slide show](http://thesecretlivesofdata.com/raft/)
  - the [quick links](https://raft.github.io/) on github.io
  - the [paper](https://raft.github.io/raft.pdf)
  - the [implementations list](https://raft.github.io/#implementations)
  
- zero-conf protocols:
  - [python-zeroconf](https://github.com/jstasiak/python-zeroconf), it's built on threads..
  
- BFT variants:
  - to be filled out by the crypto pump bois
  
  
  As always, lurkers engage.

**Comment by guilledk** [comment]
Crypto pump is here:

BFT options:

- Proof of Work: not appropriate to the use case and also highly inefficient
- Proof of Stake: more efficient but also has possible centralization issues
- Delegated Proof of Stake: even more efficient than PoS but with similar centralization issues [(video 3:04)](https://www.youtube.com/watch?v=OVKAOwzAwHI)

I'm currently working in a startup developing decentralized applications on the DPoS blockchain [Telos](https://www.telos.net/), Its a fork of the bigger [EOS](https://eos.io/) blockchain.

Some good things about it:
- Smart Contract language is C++ uses llvm backend
- Resource model is great CPU/NET is practically free as you just need to stake a bit of tokens, RAM is what you need for storing permanent information in the blockchain, but as with any token you are free to sell it at any point getting your money back or even making a profit.

Basically every time you need for example, voting or an escrow you can write your backend in a smart contract and interact with it from python.

[DOCS](https://docs.telos.net/)

**Comment by goodboy** [comment]
thanks @guilledk, got any links on performance tradeoffs by chance?

No rush either.

**Comment by gc-ss** [comment]
Hashicorp has some great, single binary (so deployment is not a challenge) solutions to:

> custom cross-host supervisors
> mult-host service architecture tooling

Hashicorp Nomad 

> we need both a service discovery system (presumably by relying on a (set of) protocols) and

Hashicorp Consul

> a modern concensus system which these days is usually in the form of either a centralized algo such as raft 

Hashicorp Consul also exposes their battle tested raft library https://github.com/hashicorp/raft

https://www.hashicorp.com/resources/everybody-talks-gossip-serf-memberlist-raft-swim-hashicorp-consul

**Comment by goodboy** [comment]
Huh, i had thought i already dumped my little knowledge base on recent research but i guess not..

here's some interesting (maybe more moden) highlights maybe worth investigating,

- https://en.wikipedia.org/wiki/Rendezvous_protocol found while spelunking,
  * https://github.com/rustdesk/rustdesk/wiki/How-does-RustDesk-work%3F

- found link from `radicle` docs, https://en.wikipedia.org/wiki/Secure_Scuttlebutt

and from a general web-search/wikipedia sprint,

- https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing
- https://spec.matrix.org/unstable/server-server-api/#room-state-resolution
- https://spec.matrix.org/unstable/rooms/v2/
- https://en.wikipedia.org/wiki/Multicast_DNS
- https://en.wikipedia.org/wiki/Name_Service_Switch
- https://en.wikipedia.org/wiki/Link-Local_Multicast_Name_Resolution
- https://en.wikipedia.org/wiki/Zero-configuration_networking#Name_service_discovery
- https://en.wikipedia.org/wiki/Service_Location_Protocol
- https://en.wikipedia.org/wiki/Peer_Name_Resolution_Protocol
- https://en.wikipedia.org/wiki/Name_Service_Switch
- https://raft.github.io/raft.pdf

(UPDATE, lul i guess i did post the final list prior here https://github.com/goodboy/tractor/issues/216#issuecomment-999157291 🫣 )
